flink-mesos/src/main/scala/org/apache/flink/runtime/clusterframework/ContaineredJobManager.scala

```diff
@@ -0,0 +1,174 @@
+  *                         [[org.apache.flink.runtime.executiongraph.ExecutionGraph]]
+  * @param instanceManager Instance manager to manage the registered
+  *                        [[org.apache.flink.runtime.taskmanager.TaskManager]]
+  * @param scheduler Scheduler to schedule Flink jobs
+  * @param libraryCacheManager Manager to manage uploaded jar files
+  * @param archive Archive for finished Flink jobs
+  * @param restartStrategyFactory Restart strategy to be used in case of a job recovery
+  * @param timeout Timeout for futures
+  * @param leaderElectionService LeaderElectionService to participate in the leader election
+  */
+abstract class ContaineredJobManager(
+                      flinkConfiguration: FlinkConfiguration,
+                      executorService: ExecutorService,
+                      instanceManager: InstanceManager,
+                      scheduler: FlinkScheduler,
+                      libraryCacheManager: BlobLibraryCacheManager,
+                      archive: ActorRef,
+                      restartStrategyFactory: RestartStrategyFactory,
+                      timeout: FiniteDuration,
+                      leaderElectionService: LeaderElectionService,
+                      submittedJobGraphs : SubmittedJobGraphStore,
+                      checkpointRecoveryFactory : CheckpointRecoveryFactory,
+                      savepointStore: SavepointStore,
+                      jobRecoveryTimeout: FiniteDuration,
+                      metricsRegistry: Option[FlinkMetricRegistry])
+  extends JobManager(
+    flinkConfiguration,
+    executorService,
+    instanceManager,
+    scheduler,
+    libraryCacheManager,
+    archive,
+    restartStrategyFactory,
+    timeout,
+    leaderElectionService,
+    submittedJobGraphs,
+    checkpointRecoveryFactory,
+    savepointStore,
+    jobRecoveryTimeout,
+    metricsRegistry) {
+
+  val jobPollingInterval: FiniteDuration
+
+  // indicates if this JM has been started in a dedicated (per-job) mode.
+  var stopWhenJobFinished: JobID = null
+
+  override def handleMessage: Receive = {
+    handleContainerMessage orElse super.handleMessage
+  }
+
+  def handleContainerMessage: Receive = {
+
+    case msg @ (_: RegisterInfoMessageListener | _: UnRegisterInfoMessageListener) =>
+      // forward to ResourceManager
+      currentResourceManager match {
+        case Some(rm) =>
+          // we forward the message
+          rm.forward(decorateMessage(msg))
+        case None =>
+        // client has to try again
+      }
+
+    case msg: ShutdownClusterAfterJob =>
+      val jobId = msg.jobId()
+      log.info(s"ApplicationMaster will shut down session when job $jobId has finished.")
+      stopWhenJobFinished = jobId
+      // trigger regular job status messages (if this is a dedicated/per-job cluster)
+      if (stopWhenJobFinished != null) {
+        context.system.scheduler.schedule(0 seconds,
```

Comment:
This is code that is cut-pasted from the `YarnJobManager`.  Indeed I'm meaning to have `YarnJobManager` extend from the above`ContaineredJobManager` to consolidate the logic.   I don't want to challenge the logic itself; @mxm might best be able to address your remark.